src/librustc/ty/query/plumbing.rs

```diff
@@ -139,39 +149,54 @@ impl<'a, 'tcx, Q: QueryDescription<'tcx>> JobOwner<'a, 'tcx, Q> {
                                 query_blocked_prof_timer = Some(tcx.prof.query_blocked());
                             }
 
-                            job.clone()
+                            // Create the id of the job we're waiting for
+                            let id = QueryJobId {
+                                job: job.id,
+                                shard: u16::try_from(shard).unwrap(),
+                                kind: Q::dep_kind(),
+                            };
+
+                            job.latch(id)
                         }
                         QueryResult::Poisoned => FatalError.raise(),
                     }
                 }
                 Entry::Vacant(entry) => {
+                    let jobs = &mut lock.jobs;
+
                     // No job entry for this query. Return a new one to be started later.
                     return tls::with_related_context(tcx, |icx| {
-                        // Create the `parent` variable before `info`. This allows LLVM
-                        // to elide the move of `info`
-                        let parent = icx.query.clone();
-                        let info = QueryInfo { span, query: Q::query(key.clone()) };
-                        let job = Lrc::new(QueryJob::new(info, parent));
-                        let owner = JobOwner { cache, job: job.clone(), key: (*key).clone() };
+                        // Generate an id unique within this shard.
+                        let id = jobs.checked_add(1).unwrap();
+                        *jobs = id;
+                        let id = QueryShardJobId(NonZeroU32::new(id).unwrap());
+
+                        let global_id = QueryJobId {
+                            job: id,
+                            shard: u16::try_from(shard).unwrap(),
+                            kind: Q::dep_kind(),
+                        };
```

Comment:
Is there are reason for not creating the `global_id` outside the closure. That seems simpler than capturing the `jobs` reference.

Reply:
No, this is was my default way of dealing with this particular borrow error, I've refactored it.